### PR TITLE
Revamp Command and Protocol API

### PR DESCRIPTION
**Pull request by pipermerriam**
### What was wrong?

Why Notes:

- All command payloads are strongly typed.
- Since payloads are our own objects they can have things like computed fields.
- No more separation between command instance and payload instance (payload is now part of the command instance)
- Given a command instance and the protocol instance it belongs to you can have a generic `Protocol.send(command)` method (no more need for individual `send_thing` methods on the protocol class.
- Convenience APIs for sending protocol methods fit nicely on the new `Application` API
- Composition over inheritance. `Command` classes no longer rely on inheritance to extend their functionality.
- Better suited for re-usable code that serves protocol command over the event bus.  We only need a single lahja event type which only needs `session`, `protocol_type`, `command` to transmit **any** protocol message over the wire.  (this was some of the primary motivation for doing this)

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by pipermerriam** [comment]
This looks to be pretty simple to do and results in our peer messages being very strongly typed in both directions (no more `Any` types or `Dict[str, Any]` types)

**Comment by pipermerriam** [comment]
This one is going well but it's ballooned in a way I wasn't expecting.  I think I'm past the worst of it now.

PR is going to end up far too big.  Once it's done and passing I'll start chipping it down into smaller pieces.

File: trinity/protocol/eth/peer.py [hunk 1]
```diff
@@ -189,6 +166,7 @@ class ETHPeerPoolEventServer(PeerPoolEventServer[ETHPeer]):
 
         await super()._run()
 
+    """
```
Comment (carver) [inline trinity/protocol/eth/peer.py]: Maybe add a #TODO to make sure it doesn't get merged like this?

File: trinity/protocol/les/peer.py [hunk 2]
```diff
@@ -195,6 +194,7 @@ def __init__(self,
 
         await super()._run()
 
+    """
```
Comment (carver) [inline trinity/protocol/les/peer.py]: Another tag to remember to clean this up.

**Comment by pipermerriam** [comment]
ugg, 3k lines of code and still growing

**Comment by pipermerriam** [comment]
Some notes:

Ran into a snag trying to upgrade the request/response over event bus API.  Since the `TResult` types from the exchange API are "rich" versions of the vanilla response payload, we can't just use a generic event.   This can be fixed if we move this normalization into the `CommandPayload` objects and get rid of the normalization API (though we still have to figure out how to offload the computationally expensive part).

**Comment by pipermerriam** [comment]
Made great progress on this today.  Got most of the tests passing.  I think I've updated all the various code paths and just have a few leftovers to cleanup, after which I can carve off a few pieces to get this into more manageable chunks.  Currently at +3.8k/-3k SLOC

File: p2p/p2p_proto.py [hunk 3]
```diff
@@ -1,154 +1,120 @@
...
                 f"Snappy support is not supported before version 5 of the p2p "
                 f"protocol.  Currently using version `{self.version}`"
             )
-        super().__init__(transport, cmd_id_offset=cmd_id_offset, snappy_support=False)
+        super().__init__(transport, command_id_offset, snappy_support=False)
 
 
-class P2PProtocol(BaseP2PProtocol):
-    version = 5
+class P2PProtocolV5(BaseP2PProtocol):
```
Comment (pipermerriam) [inline p2p/p2p_proto.py]: This rename could be extracted.

File: trinity/protocol/common/peer.py [hunk 4]
```diff
@@ -61,11 +62,16 @@ class BaseChainPeer(BasePeer):
...
+    def chain_api(self) -> Union[ETHAPI, LESV1API, LESV2API]:
         if self.connection.has_logic(ETHAPI.name):
             return self.connection.get_logic(ETHAPI.name, ETHAPI)
-        elif self.connection.has_logic(LESAPI.name):
-            return self.connection.get_logic(LESAPI.name, LESAPI)
+        elif self.connection.has_logic(LESV1API.name):
+            if self.connection.has_protocol(LESProtocolV2):
+                return self.connection.get_logic(LESV2API.name, LESV2API)
+            elif self.connection.has_protocol(LESProtocolV1):
+                return self.connection.get_logic(LESV1API.name, LESV1API)
```
Comment (pipermerriam) [inline trinity/protocol/common/peer.py]: These are a little ugly but  I don't necessarily think they have to be cleaned up now.

File: trinity/protocol/les/commands.py [hunk 5]
```diff
@@ -1,244 +1,387 @@
...
+        sedes.big_endian_int,
+    ))),
+))
+
+
+class GetProofsV1(BaseCommand[GetProofsPayload]):
+    protocol_command_id = 8
+    serialization_codec = RLPCodec(
+        sedes=GET_PROOFS_V1_STRUCTURE,
+        normalize_fn=compose(
```
Comment (pipermerriam) [inline trinity/protocol/les/commands.py]: These deserve custom functions.

File: trinity/tools/factories.py [hunk 6]
```diff
@@ -1,303 +0,0 @@
...
-
-
-class LESV2HandshakerFactory(factory.Factory):
-    class Meta:
-        model = LESV2Handshaker
-
-    handshake_params = factory.SubFactory(LESHandshakeParamsFactory, version=LESProtocolV2.version)
-
-
-class LESV1Peer(LESPeer):
```
Comment (pipermerriam) [inline trinity/tools/factories.py]: This re-organization can be in its own PR

File: trinity/protocol/eth/peer.py [hunk 7]
```diff
@@ -179,33 +174,33 @@ class ETHPeerPoolEventServer(PeerPoolEventServer[ETHPeer]):
     async def handle_block_headers_event(self, event: SendBlockHeadersEvent) -> None:
         await self.try_with_session(
             event.session,
-            lambda peer: peer.sub_proto.send_block_headers(event.headers)
+            lambda peer: peer.sub_proto.send(event.command)
```
Comment (carver) [inline trinity/protocol/eth/peer.py]: Is `sub_proto` independent of the protocol now? ie~ could you send an LES command via `peer.sub_proto.send()`?

If not, maybe this should be `peer.eth_api.send`.

**Comment by pipermerriam** [comment]
Still have failing integration tests that I think are real failures.  Also this:

```bash
   ERROR  2019-11-05 08:56:02,098           ETHPeerPool  Unexpected error during auth/p2p handshake with <Node(0xa6f0a1802760ba958a9bcefc9c4e97b7188adf63a44ae5674d08d9eacee314d20dc6347b95aaaef59ae842223aa90c7b3d5ca4a5678de65154080c48737a14ff@23.111.188.207:30303)>
Traceback (most recent call last):
  File "/home/piper/projects/trinity/p2p/peer_pool.py", line 361, in connect
    timeout=HANDSHAKE_TIMEOUT,
  File "/home/piper/projects/trinity/p2p/cancellable.py", line 20, in wait
    return await self.wait_first(awaitable, token=token, timeout=timeout)
  File "/home/piper/projects/trinity/p2p/cancellable.py", line 43, in wait_first
    return await token_chain.cancellable_wait(*awaitables, timeout=timeout)
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/cancel_token/token.py", line 178, in cancellable_wait
    return done.pop().result()
  File "/home/piper/projects/trinity/p2p/peer.py", line 490, in handshake
    token=self.cancel_token
  File "/home/piper/projects/trinity/p2p/handshake.py", line 312, in dial_out
    token=token,
  File "/home/piper/projects/trinity/p2p/handshake.py", line 222, in negotiate_protocol_handshakes
    token,
  File "/home/piper/projects/trinity/p2p/handshake.py", line 129, in _do_p2p_handshake
    async for _, cmd in stream_transport_messages(transport, base_protocol, token=token):
  File "/home/piper/projects/trinity/p2p/multiplexer.py", line 56, in stream_transport_messages
    msg = await transport.recv(token)
  File "/home/piper/projects/trinity/p2p/transport.py", line 307, in recv
    header_data = rlp.decode(padded_header[3:], sedes=HEADER_DATA_SEDES, strict=False)
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/rlp/codec.py", line 237, in decode
    obj = sedes.deserialize(item, **kwargs)
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/eth_utils/functional.py", line 45, in inner
    return callback(fn(*args, **kwargs))
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/rlp/sedes/lists.py", line 89, in deserialize
    serial)
rlp.exceptions.ListDeserializationError: Deserializing list length (1) does not match sedes (2)
   ERROR  2019-11-05 08:56:02,116           ETHPeerPool  unexpected error during peer connection
Traceback (most recent call last):
  File "/home/piper/projects/trinity/p2p/peer_pool.py", line 209, in maybe_connect_more_peers
    for backend in self.peer_backends
  File "/home/piper/projects/trinity/p2p/peer_pool.py", line 191, in _add_peers_from_backend
    await self.connect_to_nodes(iter(candidates))
  File "/home/piper/projects/trinity/p2p/peer_pool.py", line 422, in connect_to_nodes
    loop=self.get_event_loop(),
  File "/home/piper/projects/trinity/p2p/peer_pool.py", line 436, in connect_to_node
    peer = await self.connect(node)
  File "/home/piper/projects/trinity/p2p/peer_pool.py", line 361, in connect
    timeout=HANDSHAKE_TIMEOUT,
  File "/home/piper/projects/trinity/p2p/cancellable.py", line 20, in wait
    return await self.wait_first(awaitable, token=token, timeout=timeout)
  File "/home/piper/projects/trinity/p2p/cancellable.py", line 43, in wait_first
    return await token_chain.cancellable_wait(*awaitables, timeout=timeout)
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/cancel_token/token.py", line 178, in cancellable_wait
    return done.pop().result()
  File "/home/piper/projects/trinity/p2p/peer.py", line 490, in handshake
    token=self.cancel_token
  File "/home/piper/projects/trinity/p2p/handshake.py", line 312, in dial_out
    token=token,
  File "/home/piper/projects/trinity/p2p/handshake.py", line 222, in negotiate_protocol_handshakes
    token,
  File "/home/piper/projects/trinity/p2p/handshake.py", line 129, in _do_p2p_handshake
    async for _, cmd in stream_transport_messages(transport, base_protocol, token=token):
  File "/home/piper/projects/trinity/p2p/multiplexer.py", line 56, in stream_transport_messages
    msg = await transport.recv(token)
  File "/home/piper/projects/trinity/p2p/transport.py", line 307, in recv
    header_data = rlp.decode(padded_header[3:], sedes=HEADER_DATA_SEDES, strict=False)
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/rlp/codec.py", line 237, in decode
    obj = sedes.deserialize(item, **kwargs)
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/eth_utils/functional.py", line 45, in inner
    return callback(fn(*args, **kwargs))
  File "/home/piper/python-environments/trinity/lib/python3.6/site-packages/rlp/sedes/lists.py", line 89, in deserialize
    serial)
rlp.exceptions.ListDeserializationError: Deserializing list length (1) does not match sedes (2)
```

Should be a simple error to handle, just need add it.

**Comment by pipermerriam** [comment]
@carver I've run this for a number of hours beam syncing with no problems.  There is something broken with LES that I'm still trying to debug.

**Comment by pipermerriam** [comment]
@carver this should be green now.  I need to give it one more visual pass to check and see if I left anything odd in while debugging some things in the last few days.  I still need to run it for an extended time now that I fixed a number of things, but I my previous long runs with normal beam sync were stable and didn't yield any issues that haven't already been addressed.

**Comment by carver** [comment]
Okay cool, taking a look...

File: p2p/commands.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        # Frame-size is a 3-bit integer
+        header = struct.pack('>I', frame_size)[1:] + RLPX_HEADER_DATA
+        body = cmd_id_data + payload_data
+
+        return Message(header, body)
+
+    @classmethod
+    def decode(cls: Type[TCommand], message: MessageAPI, snappy_support: bool) -> TCommand:
+        if snappy_support:
+            payload_data = cls.compression_codec.decompress(message.body[1:])
```
Comment (carver) [inline p2p/commands.py]: Any way we can verify the first byte here (`command_id`) against the command we're using to decode it?

It doesn't look possible, because the offset version of the command id isn't present on the command class.

File: p2p/message.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
+from typing import Any
+
+import rlp
+
+from p2p.abc import MessageAPI
+
+
+class Message(MessageAPI):
```
Comment (carver) [inline p2p/message.py]: Maybe this class could use a `encoded_payload` property that is `self.body[1:]`, so that we don't have that slightly awkward slicing in the command decode.

File: p2p/multiplexer.py [hunk 10]
```diff
@@ -38,52 +37,57 @@
...
+                        f"No protocol found for command_id {command_id}: Available "
                         f"protocol/offsets are: {protocol_infos}"
                     )
 
-        msg_proto = cmd_id_cache[cmd_id]
-        cmd = msg_proto.cmd_by_id[cmd_id]
-        msg = cmd.decode(raw_msg)
+        msg_proto = command_id_cache[command_id]
+        command_type = msg_proto.get_command_type_for_command_id(command_id)
+        cmd = command_type.decode(msg, msg_proto.snappy_support)
```
Comment (carver) [inline p2p/multiplexer.py]: Another thought (half-baked) about snappy support. Since it's the protocol that knows about support, instead of the command, maybe the protocol should do the decompression?

Like:
```py
unpacked_message = msg_proto.decompress(msg)  # this is just a no-op if msg_proto.snappy_support=False
cmd = command_type.decode(unpacked_message)
```

File: p2p/multiplexer.py [hunk 11]
```diff
@@ -38,52 +37,57 @@
...
 
-        yield msg_proto, cmd, msg
+        logger.debug('got %s from %s', cmd.__class__.__name__, transport.session)
```
Comment (carver) [inline p2p/multiplexer.py]: I'm guessing this will need to switch to `debug2` pretty soon, so it doesn't flood the logs.

File: p2p/commands.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            payload_data = raw_payload_data
+
+        cmd_id_data = rlp.encode(cmd_id, sedes=rlp.sedes.big_endian_int)
+        frame_size = len(cmd_id_data) + len(payload_data)
+
+        if frame_size.bit_length() > 24:
+            raise ValueError("Frame size has to fit in a 3-byte integer")
+
+        # Frame-size is a 3-bit integer
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply p2p/commands.py]: I just realized I can do `int.to_bytes(3, 'big')` and remove the need to slice off the 4th byte.

File: p2p/multiplexer.py (hunk 10 above)
Comment (pipermerriam) [reply p2p/multiplexer.py]: :+1: I was arriving at this myself.  Simplifies the `Command` class and puts it in the place that it really belongs.

File: p2p/multiplexer.py (hunk 11 above)
Comment (pipermerriam) [reply p2p/multiplexer.py]: This is a leftover that should be removed from my debugging.

File: p2p/commands.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    decode_strict: bool
+
+    def __init__(self,
+                 sedes: Any,
+                 decode_strict: bool = True,
+                 process_outbound_payload_fn: Callable[[TCommandPayload], Any] = None,
+                 process_inbound_payload_fn: Callable[[Any], TCommandPayload] = None) -> None:
+        if not hasattr(sedes, 'serialize'):
+            raise TypeError("Invalid sedes: ...")
+        if not hasattr(sedes, 'deserialize'):
```
[the remark this answers is not in the document]
Comment (cburgdorf) [reply p2p/commands.py]: There's a backport (`from typing_extensions import Protocol`) that we already use at a few places.

File: trinity/protocol/eth/peer.py (hunk 7 above)
Comment (carver) [inline trinity/protocol/eth/peer.py]: `eth_api`? (plus a couple others below)

File: trinity/protocol/les/peer.py [hunk 14]
```diff
@@ -163,7 +175,7 @@ def __init__(self,
             SendBlockHeadersEvent,
             lambda ev: self.try_with_session(
                 ev.session,
-                lambda peer: peer.sub_proto.send_block_headers(ev.headers, ev.buffer_value, ev.request_id)  # noqa: E501
+                lambda peer: peer.sub_proto.send(ev.command)
```
Comment (carver) [inline trinity/protocol/les/peer.py]: Any reason we need to leave this as sub_proto? vs...

```suggestion
                lambda peer: peer.les_api.send(ev.command)
```

File: trinity/tools/factories/eth/payloads.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    network_id = MAINNET_NETWORK_ID
+    total_difficulty = 1
+    head_hash = factory.SubFactory(BlockHashFactory)
+    genesis_hash = factory.SubFactory(BlockHashFactory)
+
+    @classmethod
+    def from_headerdb(cls, headerdb: HeaderDatabaseAPI, **kwargs: Any) -> StatusPayload:
+        head = headerdb.get_canonical_head()
+        head_score = headerdb.get_score(head.hash)
+        # TODO: https://github.com/ethereum/py-evm/issues/1847
```
Comment (carver) [inline trinity/tools/factories/eth/payloads.py]: Should the linked issue be reopened?

File: tests/p2p/test_peer_subscriber.py [hunk 16]
```diff
@@ -35,11 +38,11 @@ class AllSubscriber(PeerSubscriber):
...
         peer.add_subscriber(all_subscriber)
 
-        remote.sub_proto.send_broadcast_data(b'value-a')
-        remote.sub_proto.send_broadcast_data(b'value-b')
-        remote.sub_proto.send_get_sum(7, 8)
-        remote.sub_proto.send_get_sum(1234, 4321)
-        remote.sub_proto.send_broadcast_data(b'value-b')
+        remote.sub_proto.send(BroadcastData(BroadcastDataPayload(b'value-a')))
+        remote.sub_proto.send(BroadcastData(BroadcastDataPayload(b'value-b')))
+        remote.sub_proto.send(GetSum(GetSumPayload(7, 8)))
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/p2p/test_peer_subscriber.py]: Yeah, didn't do that due to `mypy` limitations but I'm :+1: on that type of API because yes, the `Command` does know about it's payload type.

File: trinity/protocol/eth/peer.py (hunk 7 above)
Comment (pipermerriam) [reply trinity/protocol/eth/peer.py]: no, these stay as `sub_proto` since they have a fully instantiated `Command` instance and can just use the `send` API of the sub protocol.

File: trinity/tools/factories/eth/payloads.py (hunk 15 above)
Comment (pipermerriam) [reply trinity/tools/factories/eth/payloads.py]: No, removed comment as it is now outdated.

File: p2p/commands.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (carver) [reply p2p/commands.py]: Oh yeah, I forgot about that one, thanks! So we should be able to use that to type-check for these attributes instead of doing runtime checks.
